### PR TITLE
feat: Pro1 challenges, h6 & general mastery improvements

### PR DESCRIPTION
**Pull request by RDIL**
## Scope

- Add h6's professional mode challenges
- Some refactoring to the services that power all this to improve clarity.

### Remaining to-do(?)

- [X] Normal and pro1 mastery getting conflated
- [x] Progress bar at mission end is very incorrect
- [x] Can't use pro mode until signing out and signing back in once unlocking it (probably #509)

## Test Plan

I have in fact tested the code works on 2016. maybe some extra regression testing is needed for newer game versions, but hey, that's what betas are for

## Checklist

-   [x] I have run Prettier to reformat any changed files
-   [x] I have verified my changes work

### CODE REVIEW

File: components/scoreHandler.ts [hunk 1]
```diff
@@ -1144,6 +1206,10 @@ export async function getMissionEndData(
         },
     }
 
+    if (isDryRun) {
+        return result
+    }
```
Comment (RDIL) [inline components/scoreHandler.ts:1211]: Below this is only logic for committing scores & handling livesplit updates, which we can just ignore.

File: components/scoreHandler.ts [hunk 2]
```diff
@@ -387,7 +388,6 @@ export function calculateSniperScore(
 
     let timeBonus = 0
 
-    // TODO? generate this curve from contractSession.scoring.Settings["timebonus"] somehow
```
Comment (RDIL) [inline components/scoreHandler.ts:390]: Tried this, it's not possible, the curve is not linear.

File: components/2016/legacyProfileRouter.ts [hunk 3]
```diff
@@ -155,7 +147,6 @@ legacyProfileRouter.post(
                 MustBeSaved: progression.MustBeSaved,
             })
         }
-        // TODO: HELP! Please DM rdil if you see this
```
Comment (RDIL) [inline components/2016/legacyProfileRouter.ts:158]: (Vague todo I wrote for myself when realizing how big of a pain pro1 challenges would be)

File: components/candle/challengeHelpers.ts [hunk 4]
```diff
@@ -135,6 +182,7 @@ function isChallengeInContract(
     locationId: string,
```
Comment (RDIL) [inline components/candle/challengeHelpers.ts:182]: Todo: can we just pass the filter object here?

File: components/menuData.ts [hunk 5]
```diff
@@ -284,75 +284,87 @@ menuDataRouter.get(
...
+                            ChallengeName: challengeData.Name,
+                            ChallengeImageUrl: challengeData.ImageName,
+                            ChallengeDescription: challengeData.Description,
+                            XPGain: challengeData.Rewards.MasteryXP,
+                        })),
+                    Drops: [],
+                    ContractCompletionBonus: 0,
+                    GroupCompletionBonus: 0,
+                    LocationHideProgression: true,
+                    Difficulty: "normal", // FIXME: is this right?
```
Comment (RDIL) [inline components/menuData.ts:355]: No it isn’t

File: components/scoreHandler.ts [hunk 6]
```diff
@@ -529,62 +529,121 @@ export function calculateSniperScore(
...
+ * @param isDryRun When true, the function will not change the user's data or commit scores.
+ */
 export async function getMissionEndData(
     query: MissionEndRequestQuery,
     jwt: JwtData,
     gameVersion: GameVersion,
-): Promise<MissionEndError | MissionEndResult> {
+    isDryRun: boolean,
+): Promise<MissionEndResult> {
     // TODO: For this entire function, add support for 2016 difficulties
```
Comment (AnthonyFuller) [inline components/scoreHandler.ts:617]: We have this now.

File: components/scoreHandler.ts [hunk 7]
```diff
@@ -655,11 +714,19 @@ export async function getMissionEndData(
...
+        locationParentId,
+        `location ${subLocation?.Properties?.ParentLocation || levelData.Metadata.Location} not found (trying to resolve parent)`,
+    )
+
+    if (gameVersion === "h1") {
+        // h1 has a separate mastery track for pro1 and normal
+        query.masteryUnlockableId = contractData.Metadata.Difficulty ?? "normal"
+    }
+
+    if (gameVersion === "h1") {
```
Comment (AnthonyFuller) [inline components/scoreHandler.ts:727]: This is here twice?

**Review by AnthonyFuller — CHANGES_REQUESTED** [review]

File: components/scoreHandler.ts
Comment (AnthonyFuller) [inline components/scoreHandler.ts:1]: This also sends the `Progression_LevelGain` event to the client to pop mastery achievements on mission end.

Comment (RDIL) [reply components/scoreHandler.ts:1]: Ref #489

**Review by AnthonyFuller — APPROVED** [review]
LGTM, not experienced the progress bar issue in at least a double digit number of runs now.

EDIT: General Mastery Improvements 🫡

**Pull request merged**